pkg/license/license.go

```diff
@@ -0,0 +1,124 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package license
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/license"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// defaultOperatorLicenseLevel is the default license level when no operator license is installed
+	defaultOperatorLicenseLevel = "basic"
+	// licensingCfgMapName is the name of the config map used to store licensing information
+	licensingCfgMapName = "elastic-licensing"
+)
+
+// LicensingInfo represents information about the operator license including the total memory of all Elastic managed
+// components
+type LicensingInfo struct {
+	Timestamp               string `json:"timestamp"`
+	EckLicenseLevel         string `json:"eck_license_level"`
+	TotalManagedMemory      string `json:"total_managed_memory"`
+	EnterpriseResourceUnits string `json:"enterprise_resource_units"`
+}
+
+// LicensingResolver resolves the licensing information of the operator
+type LicensingResolver struct {
+	operatorNs string
+	client     k8s.Client
+}
+
+// ToInfo returns licensing information given the total memory of all Elastic managed components
+func (r LicensingResolver) ToInfo(totalMemory resource.Quantity) (LicensingInfo, error) {
+	eru := inEnterpriseResourceUnits(totalMemory)
+	memoryInGB := inGB(totalMemory)
+	licenseLevel, err := r.getOperatorLicenseLevel()
+	if err != nil {
+		return LicensingInfo{}, err
+	}
+
+	return LicensingInfo{
+		Timestamp:               time.Now().Format(time.RFC3339),
+		EckLicenseLevel:         licenseLevel,
+		TotalManagedMemory:      memoryInGB,
+		EnterpriseResourceUnits: eru,
+	}, nil
+}
+
+// Save updates or creates licensing information in a config map
+func (r LicensingResolver) Save(info LicensingInfo, operatorNs string) error {
+	data, err := info.toMap()
+	if err != nil {
+		return err
+	}
+
+	log.V(1).Info("Saving", "license_info", info)
+	cm := corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
```

Comment:
Should we maybe have a label on this so that users with multiple operators can use a label selector across namespaces to query for them?

I also wonder if it should be owned by the operator sset so that if the operator is deleted, the config map is deleted as well.

Reply:
> Should we maybe have a label on this so that users with multiple operators can use a label selector across namespaces to query for them?

Seems a good idea but which one? `common.k8s.elastic.co/type: licensing-info`

> I also wonder if it should be owned by the operator sset so that if the operator is deleted, the config map is deleted as well.

Sounds like another good idea.
We need to know the name of the elastic operator statefulset.
That doesn't seem so trivial to me.

```go
cm := corev1.ConfigMap{
	ObjectMeta: metav1.ObjectMeta{
		OwnerReferences: []metav1.OwnerReference{
			{
				APIVersion:         "apps/v1",
				Kind:               "StatefulSet",
				Name:               "elastic-operator",
				UID:                "",
				Controller:         &true,
				BlockOwnerDeletion: &true,
			},
		},
```

Reply:
>Seems a good idea but which one? common.k8s.elastic.co/type: licensing-info

That works for me. Maybe `type: elastic-usage`?

>We need to know the name of the elastic operator statefulset.

We could extract the sset name from `HOSTNAME` or if we are worried something will change that, the downward API can give us the pod name. We could leave both of these for a future PR though.

Reply:
We'd also need to retrieve StatefulSets in the operator namespace so we get its UID, which is not something we currently have RBAC permissions for.

Reply:
I added the label.
For the owner's reference, this does not seem trivial to achieve.. Considering the timing and the complexity, let's see that later.